p2p/lightchain.py

```diff
@@ -333,14 +333,33 @@ def callback(r):
         return rlp.decode(rlp_account, sedes=Account)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
-    async def get_contract_code(self, block_hash: Hash32, key: bytes) -> bytes:
+    async def get_contract_code(self, block_hash: Hash32, address: Address) -> bytes:
+        """
+        :param block_hash: find code as of the block with block_hash
+        :param address: which contract to look up
+
+        :return: bytecode of the contract, ``b''`` if no code is set
+        """
         peer = await self.get_best_peer()
         request_id = gen_request_id()
-        peer.sub_proto.send_get_contract_code(block_hash, key, request_id)
+        peer.sub_proto.send_get_contract_code(block_hash, keccak(address), request_id)
         reply = await self._wait_for_reply(request_id)
+
         if not reply['codes']:
-            return b''
-        return reply['codes'][0]
+            bytecode = b''
+        else:
+            bytecode = reply['codes'][0]
+
+        # validate bytecode against a proven account
+        account = await self.get_account(block_hash, address)
+
+        if account.code_hash == keccak(bytecode):
+            return bytecode
+        else:
+            # disconnect from this bad peer
+            await self.disconnect_peer(peer, DisconnectReason.subprotocol_error)
+            # try again with another peer
+            return await self.get_contract_code(block_hash, address)
```

Comment:
A recursive call like this could potentially send us in an infinite loop; I think it'd be preferable to raise something here and have callers deal with it whichever way they want.

Reply:
How about we just retry, but with a timeout? (a single timeout across all retries, of course)

Dealing with timeouts makes sense as a caller. Dealing with a bad faith response from some specific peer that the caller didn't pick feels awkward.

Reply:
I still think this should be handled by the caller.  This feels like an anti pattern.  This function should only be responsible for retrieval/validation,  orchestration is something that seems like it should be higher up.